### ISSUE TITLE
Capabilities call and versioning

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#2162]
I'm working on adding support for running multiple versions of the API in parallel. This is because I think the old way we used to do API upgrades (e.g. switch off openstreetmap for a few days, switch it back on again with the new version) might not be tolerated in 2019 :smile: 

I started work on this at the Karlsruhe Hack Weekend and if you want to see some WIP and get a better idea what I'm thinking about then have a look at https://github.com/gravitystorm/openstreetmap-website/commit/f574838ddabf3dc871da25ae9ccc331c2537d0c1 (specifically the config/routes.rb file).

One topic came up that stumped me. We currently have two capabilities calls, that currently return the same thing:

* [`api/capabilities`](https://www.openstreetmap.org/api/capabilities) (the 'unversioned' call)
* [`api/0.6/capabilities`](https://www.openstreetmap.org/api/0.6/capabilities) (the 'versioned' call)

A hypothetical `api/0.7/capabilities` response could indicate that it is version 0.7 (in the root `osm` element) but 0.6 is still available (in the `versions` element), for example:

```
<osm version="0.7" generator="OpenStreetMap server" ...>
  <api>
    <version minimum="0.6" maximum="0.7"/>
    <area maximum="0.25"/>
    ....
```

But my immediate quandary is - for the unversioned `api/capabilities` call - what version should be shown in the root element when multiple versions are available? `<osm version="???" ...>`

It also got me thinking - should we even have an unversioned capabilities call? What happens if we want to change that response in future? Should we instead have an `api/versions` call that just indicates supported versions, and clients can then call `api/X/capabilities` for the one they want to use? Should we remove the version information from `api/X/capabilities` since it's not really a capability of that version of the API?

Does anything (e.g. editors) actually use the current version information anyway? Do requests for capabilities generally use the versioned or unversioned calls?

Discuss!

**Comment by bhousel** [comment]
> Does anything (e.g. editors) actually use the current version information anyway? Do requests for capabilities generally use the versioned or unversioned calls?

iD currently calls `api/capabilities` (unversioned) but we should change this.

It makes more sense to me to just drop the unversioned endpoint and have separate resources for:
* `api/0.6/capabilities`
* `api/0.7/capabilities`

> Should we instead have an api/versions call that just indicates supported versions, and clients can then call api/X/capabilities for the one they want to use? 

seems fine 👍 

> Should we remove the version information from api/X/capabilities since it's not really a capability of that version of the API?

I agree it can be removed, I never even noticed it was there..

**Comment by tomhughes** [comment]
The whole point of the unversioned call was to allow clients to determine if the API was a version that they supported.

The question of how to report which versions are supported has not yet arisen because we haven't had a new version since.

**Comment by simonpoole** [comment]
We currently use the unversioned call (that's what you get for reading the documentation 1/2 :-)).

While I concur that there is a bit of a hodgepodge of values returned, changing that would clearly need a new API version, which IMHO is a non-starter. But if you do want to do that it would likely make sense to move the operational status values and the blacklist to their own calls.

**Comment by tomhughes** [comment]
Thinking about this some more I believe the theory was that is was unlikely the server would ever support more than one version at once but that clients might, so they could just request the unversioned document to find out about capabilities include what version of the API to use rather than having to probe multiple candidates.

**Comment by mmd-osm** [comment]
> Thinking about this some more I believe the theory was that is was unlikely the server would ever support more than one version at once but that clients might

Somehow this made me worry a bit, as it effectively means that we would have to cut off all old clients not supporting the latest and greatest version at a fixed point in time. Every version change will be a bing bang in the future.

Also, every editing app would have to incorporate code for the new version months ahead of time to be prepared. Then there's the user's update discipline that might further complicate things. 

My gut feeling is that this might be extremely difficult to get right, but I'm happy if you can prove me wrong.

Today, we have so many different API endpoints under the umbrella of the API 0.6 that don't really belong there. Examples could be changeset discussions, gps traces, user data, and probably also map notes. Any incompatible changes here don't really warrant a new OSM API version in a strict sense IMHO, and this further complicates overall API evolution. Maybe that's worth exploring in another issue.

**Comment by tomhughes** [comment]
Well the logic was that we were never going to have the resources to support multiple API versions simultaneously and indeed it might not even be possible if the changes are big enough.

If we add an area type say then what happens to an old API user? They just don't see the areas when they download? So people then add old style ways-as-areas that duplicate them...

**Comment by gravitystorm** [comment]
Let's not get too distracted here. The work I am doing here is precisely to support multiple API versions at the same time, so @mmd-osm there's no need to worry. And any decisions that were made in the past about not supporting multiple API versions are only of historical interest, since it's almost a decade since the last version change and a decade is a long time, even outside of OpenStreetMap.

I'm not currently considering how we would support major changes like new entities or major data model changes, since I'm not proposing anything like that and nothing like that is anywhere near being agreed, never mind coded, anyway. So we can leave that for another time.

**Comment by gravitystorm** [comment]
> We currently use the unversioned [capabilities] call ...
> ..., changing [the capabilities call] would clearly need a new API version,

Yeah, this is where it gets a bit of a logjam, since the unversioned call is, by definition, unversioned!

Can we assume the following?

* There needs to be an unversioned call of some sort, in order to establish which API versions the server offers.
* The client should then request a versioned capabilities document. This could contain different information for different API versions (e.g. if tracepoints were renamed foopoints, etc).

If so, I would conclude that:

* Everything else in the unversioned capabilities call - other than which versions are supported - is unnecessary and uncertain, since it e.g. might correspond to features that have different names in different versions.
* The version element (e.g. `<version minimum="0.6" maximum="0.6"/>`) in the versioned capabilities call is both unnecessary and uncertain, since that call is supposed to describe the capabilities of that particular version of the api
* Also, the current way of describing available versions is insufficient, since it only gives two options (maximum and minimum) and doesn't describe which, if any, intermediate versions are supported (e.g. we make some trivial mistake in 0.7, and end up with 0.6 and 0.8 available and 0.7 removed).

So my proposal would then be:

* Leave the current unversioned capabilities call as-is, for backwards compatibility, and change only the maximum attribute of the  `<version>` entity. 
* The original question (what goes here in `<osm version="???" ...>` for the unversioned call) will be left as `0.6`, again for backwards compatibility.
* The api/0.6/capabilities call will be left as-is, including the `<version>` entity, since it's not clear that changing the maximum attribute here makes any sense anyway but it's best to leave it in, for backwards compatibility.

* Create a new unversioned call (api/versions) which returns only a list of available versions and no other information. That document will itself be unversioned, e.g.

```
<osm generator="OpenStreetMap server" ...>
  <api>
    <version>0.6</version>
    <version>0.7</version>
  </api>
</osm>
```
* In the `api/0.7/capabilities` call, no `<version>` element will be included.

This way, existing API calls would be unaffected, and in future a clearer separation is made between detecting which versions are supported and the capabilities of that particular version.

**Comment by simonpoole** [comment]
@gravitystorm that looks reasonable ... but it doesn't address where the operational status and the blacklist end up in. While moving these to their own endpoints kind of makes sense, we are adding a couple of round trips more to interactions with the API then.

**Comment by gravitystorm** [comment]
> it doesn't address where the operational status and the blacklist end up in.

I would propose leaving them in the versioned capabilities calls. I wouldn't have them in an unversioned call, because the format of either of those sections could conceivably change, and that's really hard to do unless they are already versioned. I don't propose separating them into different calls, since as you say that adds more round trips.

**Comment by gravitystorm** [comment]
@don-vip Do you have any thoughts on this topic?
